rust/kernel/cred.rs

```diff
@@ -6,41 +6,6 @@
 //!
 //! Reference: <https://www.kernel.org/doc/html/latest/security/credentials.html>
 
-use crate::{bindings, AlwaysRefCounted};
-use core::cell::UnsafeCell;
+use crate::types::unsafe_define_aref_type;
 
-/// Wraps the kernel's `struct cred`.
-///
-/// # Invariants
-///
-/// Instances of this type are always ref-counted, that is, a call to `get_cred` ensures that the
-/// allocation remains valid at least until the matching call to `put_cred`.
-#[repr(transparent)]
-pub struct Credential(pub(crate) UnsafeCell<bindings::cred>);
-
-impl Credential {
-    /// Creates a reference to a [`Credential`] from a valid pointer.
-    ///
-    /// # Safety
-    ///
-    /// The caller must ensure that `ptr` is valid and remains valid for the lifetime of the
-    /// returned [`Credential`] reference.
-    pub(crate) unsafe fn from_ptr<'a>(ptr: *const bindings::cred) -> &'a Self {
-        // SAFETY: The safety requirements guarantee the validity of the dereference, while the
-        // `Credential` type being transparent makes the cast ok.
-        unsafe { &*ptr.cast() }
-    }
-}
-
-// SAFETY: The type invariants guarantee that `Credential` is always ref-counted.
-unsafe impl AlwaysRefCounted for Credential {
-    fn inc_ref(&self) {
-        // SAFETY: The existence of a shared reference means that the refcount is nonzero.
-        unsafe { bindings::get_cred(self.0.get()) };
-    }
-
-    unsafe fn dec_ref(obj: core::ptr::NonNull<Self>) {
-        // SAFETY: The safety requirements guarantee that the refcount is nonzero.
-        unsafe { bindings::put_cred(obj.cast().as_ptr()) };
-    }
-}
+unsafe_define_aref_type!(pub struct Credential(cred); get = get_cred; put = put_cred;);
```

Comment:
Since this is an `unsafe` macro, we probably would need to justify its invocation, right?